src/script/interpreter.cpp

```diff
@@ -606,10 +606,29 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     // To provide for future soft-fork extensibility, if the
                     // operand has the disabled lock-time flag set,
                     // CHECKSEQUENCEVERIFY behaves as a NOP.
-                    if ((nSequence & CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0)
+                    // This also covers CTxIn::SEQUENCE_FINAL since all bits are
+                    // set in that case.
+                    if ((nSequence & CTxIn::SEQUENCE_LOCKTIME_DISABLE_FLAG) != 0) {
+                        // CHECKSEQUENCEVERIFY behaving as NOP, set error appropriately
+                        if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
+                            return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                         break;
+                    }
+
+                    // Only when discouraging un-upgraded semantics,
+                    // If uninterpreted fields are set, treat and reject as a NOP.
+                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS &&
+                        (nSequence & ~(CTxIn::SEQUENCE_LOCKTIME_TYPE_FLAG | CTxIn::SEQUENCE_LOCKTIME_MASK)) != 0) {
```

Comment:
I wonder if this second discouraging check is safe w.r.t to propagation of CSV'ed inputs ?

If the spending wallet has decided to explicitly signal RBF with a range of bits in the uninterpreted fields, this transaction would be rejected as non-standard ? Note, such weird range of bits could be a protocol watermark such as the Lightning obscured commitment number in commitment transactions.

Further, I've a second wonder about this discouraging effectiveness w.r.t to hypothetical future semantic. If we assume the upgrade semantics to be well-designed, they will set bit31 to disable bip69 semantic. Doing so, if they need more than the 14 uninterpreted bits, they might reuse the interpreted ones to encode their new field ? If they do so, the SEQUENCE_LOCKTIME_MASK bits might be set and not catch by this discouraging check.

I think those wonders express a trade-off, if we make this discouraging check more efficient to better preserve upgradeability, we're increasing the risk to silently the broadcast of any Bitcoin software interpreting `nSequence` for a special purpose ?

Reply:
Well, any time CSV is enabled, RBF is enabled, right? So I'm not sure why they'd add something in the uninterpreted section for that.

I'm not concerned with the future upgrading checks here. We'll simply need to write new policy/standard rules in the appropriate places to permit those sequences to be used. In the future, we don't want any sequence in our mempool we don't understand because that might make us (if un-upgraded) mine a bad block.

If someone is using nSequence for a special purpose they should write a BIP to document it or otherwise communicate the use (e.g., via writing tests and test vectors for core to comply with). Otherwise there are many ways we might accidentally break someones weird software.